### PR TITLE
Refactor compression code to expand URL support

### LINKED ISSUE #12688: CLN: consolidate compression inference as much as possible
we have 3 copies of the compression inference routines in `parser.py, parser.py, and io/common.py`. See how much we can reduce this (_some_ is necessary as sometimes we need to actually return a file handle or need to read in and re-wrap).

xref #12175

### LINKED ISSUE #14570: Reading bz2, zip, and xz-compressed files from URL fails
Previously, we've discussed reading tables form gzip-compressed URLs (see https://github.com/pandas-dev/pandas/issues/8685 and https://github.com/pandas-dev/pandas/pull/10649). This is an essential feature as data storage increasingly migrates to the cloud.

I now realize that the previous work only added support for gzipped URLs and not bz2, zip, or xz compression. However, [the documentation](http://pandas.pydata.org/pandas-docs/version/0.19.0/generated/pandas.read_csv.html) doesn't reflect this. See [this notebook](https://github.com/dhimmel/iris/blob/1418a9abe7608c3c1e821138c120e05aa8116ecb/pandas-bug.ipynb) to see the errors that are generated.

So :+1: for allowing URL locations for all supported compression formats.

### PR DESCRIPTION
**Pull request by dhimmel**
Work in progress. Picks up where #13340 left off.
 
closes #12688 
closes #13340 
closes #14570

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -276,7 +276,7 @@ def file_path_to_url(path):
 
 # ZipFile is not a context manager for <= 2.6
 # must be tuple index here since 2.6 doesn't use namedtuple for version_info
-if sys.version_info[1] <= 6:
+if compat.PY2 and sys.version_info[1] <= 6:
```
Comment (jorisvandenbossche) [inline pandas/io/common.py]: We don't support python 2.6 anymore, so I think this check can be removed

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -1791,20 +1791,10 @@ def __init__(self, f, **kwds):
...
-        # in Python 3, convert BytesIO or fileobjects passed with an encoding
-        elif compat.PY3 and isinstance(f, compat.BytesIO):
-            from io import TextIOWrapper
-
-            f = TextIOWrapper(f, encoding=self.encoding)
-            self.handles.append(f)
+        f = _get_handle(f, 'r', encoding=self.encoding,
+                        compression=self.compression,
+                        memory_map=self.memory_map)
+        self.handles.append(f)
```
Comment (jorisvandenbossche) [inline pandas/io/parsers.py]: This line (adding each `f` to ``handles``) will give problems, as only the handles we opened ourselves should be added, not handles passed by the user (previously those were only added inside the if statements)

**Review by jorisvandenbossche — COMMENTED** [review]
Can you add tests for the new features you added? (the compressed sources from urls)

File: pandas/io/parsers.py (hunk 2 above)
Comment (jorisvandenbossche) [inline pandas/io/parsers.py]: You can see the effect in the failure of the test_file_handles test

File: pandas/io/common.py [hunk 3]
```diff
@@ -285,53 +285,84 @@ def ZipFile(*args, **kwargs):
     ZipFile = zipfile.ZipFile
 
 
-def _get_handle(path, mode, encoding=None, compression=None, memory_map=False):
-    """Gets file handle for given path and mode.
+def _get_handle(source, mode, encoding=None, compression=None,
```
Comment (jreback) [inline pandas/io/common.py]: call this ``path_or_buf`` in-line with spellings elsewhere

File: pandas/io/common.py (hunk 1 above)
Comment (dhimmel) [inline pandas/io/common.py]: :+1: removed in 28f19f193dd6d1a924ad0e14996e47223df87738.

File: pandas/io/common.py (hunk 3 above)
Comment (dhimmel) [inline pandas/io/common.py]: Addressed in 673bde0503a854c9cce64ea98c4ac28d5071996f

File: pandas/io/parsers.py (hunk 2 above)
Comment (dhimmel) [inline pandas/io/parsers.py]: @jorisvandenbossche I attempted a fix in 2f43c39e16b844a0f5552c407225387135d38e18, but it's not very elegant -- logic is repeated in ` _get_handle` and `PythonParser.__init__`.

What do you think is the best solution? What situation do we not end up opening a handle ourselves? Is it a problem that in PY3 we pass `path_or_buf` through `io.TextIOWrapper` before adding it to `handles`?

**Comment by dhimmel** [comment]
> Can you add tests for the new features you added? (the compressed sources from urls)

@jorisvandenbossche, do you want me to add xz/zip/bz2 compressed versions of `salary.table.csv` to [`pandas/io/tests/parser/data`](https://github.com/pandas-dev/pandas/tree/e1cdc4b83d7e970e9683056be722d5fa2a00fa70/pandas/io/tests/parser/data)? Also should I add my tests to [`pandas/io/tests/parser/test_network.py#L17-L35`](https://github.com/pandas-dev/pandas/blob/e1cdc4b83d7e970e9683056be722d5fa2a00fa70/pandas/io/tests/parser/test_network.py#L17-L35)?

**Comment by dhimmel** [comment]
Are we also going to have to do something to [`common.py#L239-L244`](https://github.com/pandas-dev/pandas/blob/e1cdc4b83d7e970e9683056be722d5fa2a00fa70/pandas/io/common.py#L239-L244)?

**Comment by jreback** [comment]
@dhimmel yes that should be more general

File: pandas/io/tests/parser/test_network.py [hunk 4]
```diff
@@ -27,51 +34,15 @@ def setUp(self):
...
-        url = self.base_url + '.xz'
-        url_table = read_table(url, compression="zip", engine="python")
-        tm.assert_frame_equal(url_table, self.local_table)
-
-    @tm.network
-    def test_url_zip(self):
-        url = self.base_url + '.zip'
-        url_table = read_table(url, compression="zip", engine="python")
-        tm.assert_frame_equal(url_table, self.local_table)
+    def test_compressed_urls(self):
```
Comment (dhimmel) [inline pandas/io/tests/parser/test_network.py]: @jorisvandenbossche, to try and reduce the repetition of the testing code, I made this commit (28373f1ecc0d3190dc18d98a86849ff9bbb0b841) which uses [nose test generators](http://nose.readthedocs.io/en/latest/writing_tests.html#test-generators). However, the nose doc states:

> Please note that method generators are not supported in unittest.TestCase subclasses.

As a result, the tests yielded by `test_compressed_urls` aren't actually running. I'm attempting to run the test using:

```sh
nosetests --verbosity=3 pandas/io/tests/parser/test_network.py:TestCompressedUrl.test_compressed_urls
```

Do you know the best way to proceed? I'm new to nose and pretty new to testing, so any advice will be appreciated.

Comment (jreback) [inline pandas/io/tests/parser/test_network.py]: we run yielded test in computation and io/pickle

will be much easier when we switch to pytest

Comment (dhimmel) [inline pandas/io/tests/parser/test_network.py]: Cool, I see the test generator examples in [`io/tests/test_pickle.py`](https://github.com/pandas-dev/pandas/blob/06b35db1d88e75a9e3f183b58678589cf7715381/pandas/io/tests/test_pickle.py#L194), [`computation/tests/test_compat.py`](https://github.com/pandas-dev/pandas/blob/06b35db1d88e75a9e3f183b58678589cf7715381/pandas/computation/tests/test_compat.py#L40), and [`computation/tests/test_eval.py`](https://github.com/pandas-dev/pandas/blob/06b35db1d88e75a9e3f183b58678589cf7715381/pandas/computation/tests/test_eval.py#L809). Those examples don't have a `setUp` method nor do they use the `pandas.util.testing.network` decorator. So I did my best in 0fcb3c328fcec8b1beb1ed8c968ed4c7835e5da1. It seems that one side effect is that the failure messages are less informative?

File: pandas/io/common.py [hunk 5]
```diff
@@ -274,64 +273,105 @@ def file_path_to_url(path):
...
+
+    f = path_or_buf
+    is_path = isinstance(path_or_buf, compat.string_types)
+
+    # in Python 3, convert BytesIO or fileobjects passed with an encoding
+    if compat.PY3 and isinstance(path_or_buf, compat.BytesIO):
+        from io import TextIOWrapper
+        return TextIOWrapper(path_or_buf, encoding=encoding)
+
+    elif compression:
```
Comment (jreback) [inline pandas/io/common.py]: can this be ``infer`` here? (IIRC we have an inference function based on the file ext). Is this guaranteed to be a valid compression? (its fine, though I think easy to handle it here anyhow).

File: pandas/io/tests/parser/test_network.py [hunk 6]
```diff
@@ -14,24 +14,30 @@
...
-        url = ('https://raw.github.com/pandas-dev/pandas/'
-               'master/pandas/io/tests/parser/data/salaries.csv.gz')
-        url_table = read_table(url, compression="gzip", engine="python")
-        tm.assert_frame_equal(url_table, self.local_table)
-
-    @tm.network
-    def test_url_gz_infer(self):
-        url = 'https://s3.amazonaws.com/pandas-test/salary.table.gz'
-        url_table = read_table(url, compression="infer", engine="python")
+class TestCompressedUrl:
```
Comment (jreback) [inline pandas/io/tests/parser/test_network.py]: inherit from ``object``

**Comment by jreback** [comment]
ideally if you can use some of this code in the parser routines would be great (may not be completely possible, but want to consolidate some of that duplication)

File: pandas/io/common.py (hunk 5 above)
Comment (dhimmel) [inline pandas/io/common.py]: Currently, this function (`_get_handle`) cannot consume `compression=infer`. However, it would be an easy addition:

```python
if compression == 'infer':
    assert is_path
    compression = infer_compression(path_or_buff)
```

One issue I'm having is that `_get_handle` was undocumented before this PR. So it's a bit hard for me to know what to implement. @jreback or @jorisvandenbossche: can you comment on the tentative docstring and fill in the missing details, so I know exactly what to implement?

Comment (jorisvandenbossche) [inline pandas/io/common.py]: @dhimmel What is still missing in your current docstring?

Adding `compression='infer'` here only makes sense if it would be used of course. Where happens the inference of the compression now? Because maybe it is already inferred before it is passed to this function?

Comment (dhimmel) [inline pandas/io/common.py]: > Where happens the inference of the compression now?

@jorisvandenbossche, inferrence currently happens in [`pandas/io/parsers.py`](https://github.com/pandas-dev/pandas/blob/c045e1d6774aaa32ce13def79901f1a1ad8792bf/pandas/io/parsers.py#L357-L370).

I don't think `compression == 'infer'` will ever make it this far with the current design. @jreback does this answer your question?

**Comment by dhimmel** [comment]
Regarding [`common.py#L239-L244`](https://github.com/pandas-dev/pandas/blob/e1cdc4b83d7e970e9683056be722d5fa2a00fa70/pandas/io/common.py#L239-L244), I think there's a problem. Currently, when `compression="infer"` and the `path_or_buff` is a URL, `get_filepath_or_buffer` is using the [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) header of the request to infer the compression type. However, I don't actually think compressed files on the network will provide a Content-Encoding header of their compression type. Instead I think we want to infer the compression of a URL from it's name, just as we do with paths.

For example, `https://raw.githubusercontent.com/pandas-dev/pandas/master/pandas/io/tests/parser/data/salaries.csv.bz2` would infer `bz2` compression because it ends with `.bz2`. @jreback and @jorisvandenbossche, agree?

**Comment by jreback** [comment]
i think inferring based on the file name is usually enough - but u can also use the Content if it's available

File: pandas/io/common.py (hunk 5 above)
Comment (jreback) [inline pandas/io/common.py]: @dhimmel that's fine. I just want to validate somewhere that ``compression in [None, ......]`` or raise a ValueError

File: pandas/io/common.py [hunk 7]
```diff
@@ -274,64 +273,105 @@ def file_path_to_url(path):
...
+                                 ' Only one file per ZIP: {}'
                                  .format(zip_names))
+
+        # XZ Compression
         elif compression == 'xz':
             lzma = compat.import_lzma()
-            f = lzma.LZMAFile(path, mode)
+            f = lzma.LZMAFile(path_or_buf, mode)
+
+        # Unrecognized Compression
```
Comment (jreback) [inline pandas/io/common.py]: ahh ok we are doing the validation, nvm. then. (assume we have some tests that hit this path)

File: pandas/io/parsers.py [hunk 8]
```diff
@@ -1791,19 +1791,17 @@ def __init__(self, f, **kwds):
...
-                            compression=self.compression,
-                            memory_map=self.memory_map)
-            self.handles.append(f)
-        elif self.compression:
-            f = _wrap_compressed(f, self.compression, self.encoding)
-            self.handles.append(f)
-        # in Python 3, convert BytesIO or fileobjects passed with an encoding
-        elif compat.PY3 and isinstance(f, compat.BytesIO):
-            from io import TextIOWrapper
+        add_handle = (
```
Comment (jreback) [inline pandas/io/parsers.py]: *maybe* this could be a utility fnction in ``io/common`` (unless its not used anywhere else)

File: pandas/io/tests/parser/test_network.py [hunk 9]
```diff
@@ -7,31 +7,44 @@
...
-    def test_url_gz_infer(self):
-        url = 'https://s3.amazonaws.com/pandas-test/salary.table.gz'
-        url_table = read_table(url, compression="infer", engine="python")
+class TestCompressedUrl(object):
+    
+    compression_to_extension = {
+        'gzip': '.gz',
+        'bz2': '.bz2',
+        'zip': '.zip',
+        'xz': '.xz',
```
Comment (jreback) [inline pandas/io/tests/parser/test_network.py]: is there a test for invalid compression specified?

File: pandas/io/parsers.py (hunk 8 above)
Comment (dhimmel) [inline pandas/io/parsers.py]: I think `add_handle` is only used here -- I created it for this PR, so nowhere else depends on it.

I'm still a little unsure about the purpose of `self.handles` -- should it only contain file handles because right now we're adding handles that aren't files (for example in the case of `compat.PY3 and isinstance(f, compat.BytesIO)`)?

File: pandas/io/tests/parser/test_network.py (hunk 9 above)
Comment (dhimmel) [inline pandas/io/tests/parser/test_network.py]: Did you want a specific test for invalid compression specified for a URL? As of the latest commit (bff0f3e2cdabfe02cfcbe95667a1a2ea1481fe1e), compression inference has been consolidated and invalid compression [raises an error](https://github.com/dhimmel/pandas/blob/bff0f3e2cdabfe02cfcbe95667a1a2ea1481fe1e/pandas/io/parsers.py#L374-L375) before compression is attempted.

File: pandas/io/common.py [hunk 10]
```diff
@@ -233,20 +199,16 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
...
-        to_return = (list(maybe_read_encoded_stream(req, encoding,
-                                                    compression)) +
-                     [compression])
-        return tuple(to_return)
+        url = str(filepath_or_buffer)
+        req = _urlopen(url)
+        content_encoding = req.headers.get('Content-Encoding', None)
+        if content_encoding == 'gzip':
+            # Override compression based on Content-Encoding header
+            compression = 'gzip'
```
Comment (dhimmel) [inline pandas/io/common.py]: > i think inferring based on the file name is usually enough - but u can also use the Content if it's available

@jreback, `Content-Encoding` is not a versatile way to infer compression -- `gzip` is the only compression encoded that pandas and the Content-Encoding specification both support. I've modified it so inference on a URL uses its extension. However, if `Content-Encoding` is `gzip`, then compression is overridden. I don't expect this situation to arise very often. Does this make sense or should we just totally ignore `Content-Encoding`?

**Comment by dhimmel** [comment]
I'm getting [failing builds](https://ci.appveyor.com/project/jreback/pandas-465/build/1.0.2307/job/1f7g05x811genspf) in Python 2 due to `ImportError: No module named backports`. This occurs in `compat/__init__.py` at:

```python
    def import_lzma():
        """ import the backported lzma library
        or raise ImportError if not available """
        from backports import lzma
        return lzma
```

Any idea why `backports` is not available? How should I proceed?

File: pandas/io/common.py (hunk 7 above)
Comment (dhimmel) [inline pandas/io/common.py]: > assume we have some tests that hit this path

Actually, I don't think there were any tests for invalid compression (regardless of URL or not), so I added one. https://github.com/pandas-dev/pandas/pull/14576/commits/f2ce8f85b9a7d3d1000be9b3c09310aab67e38c6

File: pandas/io/common.py (hunk 10 above)
Comment (jreback) [inline pandas/io/common.py]: yeah I agree with your assements. Since this *was* here, we can leave it. But I agree the filename inference is better / more standard.

File: pandas/io/common.py [hunk 11]
```diff
@@ -276,64 +237,105 @@ def file_path_to_url(path):
...
+        # BZ Compression
         elif compression == 'bz2':
             import bz2
-            f = bz2.BZ2File(path, mode)
+            if is_path:
+                f = bz2.BZ2File(path_or_buf, mode)
+            elif compat.PY2:
+                # Python 2's bz2 module can't take file objects, so have to
+                # run through decompress manually
+                f = StringIO(bz2.decompress(path_or_buf.read()))
```
Comment (jreback) [inline pandas/io/common.py]: I think you need to close the original file.

File: pandas/io/common.py [hunk 12]
```diff
@@ -276,64 +237,105 @@ def file_path_to_url(path):
...
+    elif is_path:
+        if compat.PY2:
+            # Python 2
+            f = open(path_or_buf, mode)
+        elif encoding:
+            # Python 3 and encoding
+            f = open(path_or_buf, mode, encoding=encoding)
         else:
-            f = open(path, mode)
+            # Python 3 and no explicit encoding
```
Comment (jreback) [inline pandas/io/common.py]: I don't think you need all of this, just

``open(path_or_buf, mode, encoding=encoding)`` 

``encoding=None`` is valid on PY2.

we don't want to suppress issue with encoding, so we don't open with ``errors``

Comment (jreback) [inline pandas/io/common.py]: oh, I c, we actually had this code from before...hmm. ok then.

File: pandas/io/common.py [hunk 13]
```diff
@@ -276,64 +237,105 @@ def file_path_to_url(path):
...
         elif compression == 'xz':
             lzma = compat.import_lzma()
-            f = lzma.LZMAFile(path, mode)
+            f = lzma.LZMAFile(path_or_buf, mode)
+
+        # Unrecognized Compression
         else:
-            raise ValueError('Unrecognized compression type: %s' %
-                             compression)
+            msg = 'Unrecognized compression: {}'.format(compression)
```
Comment (jreback) [inline pandas/io/common.py]: I would preserve the same message (or even better expand it)

File: pandas/io/parsers.py [hunk 14]
```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
...
+    is None. Otherwise, inference is perfomed using the extension of the
+    filename or URL.
+    """
+    if not isinstance(filepath_or_buffer, compat.string_types):
+        return None
+    filepath = str(filepath_or_buffer)
+    for compression, extension in _compression_to_extension.items():
+        if filepath.endswith(extension):
+            return compression
+    return None
```
Comment (jreback) [inline pandas/io/parsers.py]: I would make this raise a nice ValueError if you have an *invalid* compress (accepting ``None/infer`` as valid). , then we can just call this.

File: pandas/io/parsers.py [hunk 15]
```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
...
-                inferred_compression = 'zip'
-            elif filepath_or_buffer.endswith('.xz'):
-                inferred_compression = 'xz'
-            else:
-                inferred_compression = None
-        else:
-            inferred_compression = None
+    compression = kwds.get('compression')
+    if compression not in set(_compression_to_extension) | {None, 'infer'}:
+        msg = 'Unrecognized compression type: {}'.format(compression)
```
Comment (jreback) [inline pandas/io/parsers.py]: you can do all of this in ``_infer_compression`` (which will raise if its a bad compression method), otherwise return the compression if ``infer`` (and just return if its None)

File: pandas/io/parsers.py (hunk 8 above)
Comment (jreback) [inline pandas/io/parsers.py]: the ``self.handles`` is to allow things to be closed if WE (meaning pandas opened) them, e.g. when you wrap with ``TextWrapper`` we NEED to close them. Otherwise we try to preserve open files handles (that were passed to us)

File: pandas/io/s3.py [hunk 16]
```diff
@@ -99,9 +99,7 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
         conn = boto.connect_s3(host=s3_host, anon=True)
 
     b = conn.get_bucket(parsed_url.netloc, validate=False)
-    if compat.PY2 and (compression == 'gzip' or
-                       (compression == 'infer' and
-                        filepath_or_buffer.endswith(".gz"))):
+    if compat.PY2 and compression:
```
Comment (jreback) [inline pandas/io/s3.py]: so you can just call ``_infer_compression`` I think in most of these cases

File: pandas/io/parsers.py (hunk 8 above)
Comment (jorisvandenbossche) [inline pandas/io/parsers.py]: Another way to do this would be to handle this inside `_get_handle` and return from that function `handles` that should be added. IMO that would be more maintainable, as now this boolean expression here depends on the content of the `_get_handle` function, so better to do it there.

File: pandas/io/tests/parser/test_network.py [hunk 17]
```diff
@@ -7,31 +7,48 @@
...
+
+    def test_compressed_urls(self):
+        """Test reading compressed tables from URL."""
+        msg = ('Test reading {}-compressed tables from URL: '
+               'compression="{}", engine="{}"')
+
+        for compression, extension in self.compression_to_extension.items():
+            url = self.base_url + extension
+            # args is a (compression, engine) tuple
+            for args in [(compression, 'python'), ('infer', 'python')]:
```
Comment (jorisvandenbossche) [inline pandas/io/tests/parser/test_network.py]: Is there a reason this is only tested for the python engine?

File: pandas/io/parsers.py (hunk 15 above)
Comment (jorisvandenbossche) [inline pandas/io/parsers.py]: To improve the error message, it would also be nice to list the possible values.

File: pandas/io/tests/parser/test_network.py (hunk 17 above)
Comment (dhimmel) [inline pandas/io/tests/parser/test_network.py]: I didn't think any of the changes I was making affected the `c` engine, since I didn't change any C code. But does the C engine kick in only after the file handle is opened? If so that's great and I'll add the C engine to the test. I thought someone was going to have to re-implement this feature changes in C.

Comment (jorisvandenbossche) [inline pandas/io/tests/parser/test_network.py]: Not sure about that. Let's see what the test runs says, but if there are problems with the C engine, let's leave that for another PR. 
In any case, if those pass for the C engine, you can also use `self.read_csv` instead of `read_csv` and then the tests will automatically be run for both the python and C engine without needing to specify it explicitly

**Comment by dhimmel** [comment]
@jorisvandenbossche, I added the `c` engine tests in https://github.com/pandas-dev/pandas/pull/14576/commits/e78d022949cd610276d83a820c45768840c25d0a. But we're gettings errors in the following [tests](https://travis-ci.org/pandas-dev/pandas/jobs/181193644):

```
ERROR: Test reading bz2-compressed tables from URL: compression="bz2", engine="c"
ERROR: Test reading bz2-compressed tables from URL: compression="infer", engine="c"
```

due to `ValueError: Python 2 cannot read bz2 from open file handle`. This error is coming from [`parser.pyx#L615-L621`](https://github.com/pandas-dev/pandas/blob/b3dd9bafbfcbac1a2b6ab756e588d6e167f90439/pandas/parser.pyx#L615-L621). This code looks to duplicate functionality of `_get_handle`, but is maybe Cython code? Anyways, I've never worked with `.pyx` code. Suggestions? Should we save `c` engine for a future PR?

File: pandas/io/parsers.py [hunk 18]
```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
     return nrows
 
 
+_compression_to_extension = {
```
Comment (jreback) [inline pandas/io/parsers.py]: you can't sort None with strings in python 3.

```
 In [3]: set(['gz', 'xz']) | { None, 'infer'}
Out[3]: {None, 'xz', 'gz', 'infer'}

In [4]: sorted(set(['gz', 'xz']) | { None, 'infer'})
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-2dd21294aad8> in <module>()
----> 1 sorted(set(['gz', 'xz']) | { None, 'infer'})

TypeError: unorderable types: str() < NoneType()
```

instead you need to compare None separately (you can actually stringify things and it will work, e.g. None -> 'None') but that is non-pythonic

File: pandas/io/parsers.py [hunk 19]
```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
...
+}
+
+
+def _infer_compression(filepath_or_buffer):
+    """
+    Infer compression of a filepath or buffer. In case of buffer, compression
+    is None. Otherwise, inference is perfomed using the extension of the
+    filename or URL.
+    """
+    if not isinstance(filepath_or_buffer, compat.string_types):
```
Comment (jreback) [inline pandas/io/parsers.py]: an FYI this would exclude pathlib type objects (but can deal with that later)

File: pandas/io/parsers.py (hunk 18 above)
Comment (dhimmel) [inline pandas/io/parsers.py]: Ahh. Okay fixed that but they're are these `pandas.io.common.ParserError`s that are killing me. https://github.com/dhimmel/pandas/commit/005b673eb924c30b84484331e36d23250b76dad9#commitcomment-20080291

Comment (jreback) [inline pandas/io/parsers.py]: you prob need to rebuild the extensions. That is a relatively recent change.

**Comment by jorisvandenbossche** [comment]
Yes, let's leave issues with the C engine for another PR (also the pathlib stuff, there is an issue for it), and lets try to finish this PR.

File: pandas/io/parsers.py (hunk 18 above)
Comment (dhimmel) [inline pandas/io/parsers.py]: > you prob need to rebuild the extensions.

`python setup.py build_ext --inplace` did the trick!

https://github.com/pandas-dev/pandas/pull/14576/commits/a821b8207ea3a8ed54b3a79a7ae208da2adefc44 moves `_infer_compression` to `io/common.py`.

**Comment by dhimmel** [comment]
Still haven't migrated the handle tracking logic to `_infer_compression` as @jorisvandenbossche. The main thing holding me up is that I don't know exactly what should be considered a handle. Is this a prerequisite to merging this PR, or should we address later? 

> Yes, let's leave issues with the C engine for another PR

Okay c engine tests removed in https://github.com/pandas-dev/pandas/pull/14576/commits/bc9304389996496d92a72789c86b95b328e7925b. Prior to removing, the c-engine [tests](https://travis-ci.org/pandas-dev/pandas/builds/181502054) were all passing in 3. In 2.7 with the c-engine, the `bz2` reading from URL fails, but the other compression methods work.

I'd consider future work to add more S3 tests for all compression types. Also consolidating the [c-engine code](https://github.com/pandas-dev/pandas/blob/b3dd9bafbfcbac1a2b6ab756e588d6e167f90439/pandas/parser.pyx#L615-L621) that's making bz2 fail in 2.

When this PR is merged, it's important to me to have my authorship preserved in the commit history. I bring this up, since the first commit in this PR is from @lababidi -- I'm assuming they also care that their authorship is preserved. So just let me know if you'd like me to squash this thing in a certain way / how we should proceed.

Thanks for seeing this through. Can't wait to use the feature going forward and happy to keep working on these issues as needed.

File: pandas/io/common.py (hunk 11 above)
Comment (dhimmel) [inline pandas/io/common.py]: By following with `path_or_buf.close()`?

**Comment by jorisvandenbossche** [comment]
> The main thing holding me up is that I don't know exactly what should be considered a handle.

The handles we need to track, are all the handles (open files or file like objects) that pandas opens *itself* (not file handles provided by the user). You can check from the current `add_handle` boolean expression which ones it should be (those checks correspond to parts in the code of `_get_handle`. Tracking those opened handles inside `_get_handle` will make it a lot clearer what is going on.

I would like to see this fixed before merging

> I'd consider future work to add more S3 tests for all compression types. Also consolidating the c-engine code that's making bz2 fail in 2.

Can you open issues for those to keep track of them.

> When this PR is merged, it's important to me to have my authorship preserved in the commit history.

That's not problem, for sure! We can just squash your additional commits into one commit, and then commit those two commits of two authors to master.

File: pandas/io/common.py (hunk 11 above)
Comment (jorisvandenbossche) [inline pandas/io/common.py]: As this is a user provided file object, we shouldn't close it? @jreback or is this a special case because of the decompressing?

**Comment by dhimmel** [comment]
Will get around to finishing this early next week. Currently traveling.

**Comment by jreback** [comment]
can you rebase, problem with appeveyor which i just fixed

File: pandas/io/common.py [hunk 20]
```diff
@@ -276,64 +237,144 @@ def file_path_to_url(path):
...
         else:
-            raise ValueError('Unrecognized compression type: %s' %
-                             compression)
+            msg = 'Unrecognized compression type: {}'.format(compression)
+            raise ValueError(msg)
+
+        # In Python 3
         if compat.PY3:
             from io import TextIOWrapper
             f = TextIOWrapper(f, encoding=encoding)
```
Comment (dhimmel) [inline pandas/io/common.py]: So it looks like TextIOWrappers shouldn't count as distinct file objects. The following code closes `read_file` by `text.close()`.

```python
read_file = open(path, 'rb')
text = io.TextIOWrapper(read_file)
text.close()
```

**Comment by dhimmel** [comment]
Assuming the tests pass for https://github.com/pandas-dev/pandas/pull/14576/commits/640dc1b0ed88bfdb9164083077d17af6a87ba826, I don't think there are any more unaddressed comments.

> can you rebase, problem with appeveyor which i just fixed

@jreback rebased with appveyor now passing.

> Can you open issues for those to keep track of them.

@jorisvandenbossche: I created #14874 and #14875.

> We can just squash your additional commits into one commit, and then commit those two commits of two authors to master.

@jorisvandenbossche do you want me to squash all of my commits? If so, what should go in the commit message? Should I add the [issues we're closing](https://github.com/pandas-dev/pandas/pull/14576#issue-187096549) to the commit message?

File: pandas/io/common.py (hunk 20 above)
Comment (dhimmel) [inline pandas/io/common.py]: I addressed this by the following approach. If the object supports `.close()`, I considered it file-like and added it to `handles`. By returning a list (`handles`) containing new handles opened by `_get_handle`, we can return multiple new file objects. While you only need to close a single of these file objects to close them all, I assumed there's no harm in adding more handles than less.

**Comment by jreback** [comment]
@dhimmel you can squash or it will get squashed on the merge. in this case if you'd leave the original commit, and squash all of our subsequent ones to a single (or possibly) multiple, logically oriented commits would be great. We could in this case cherry-pick them (rather than completely squashing), to preserver authoriship (as I think when we squash ALL then authoriship is the last one). if we only have 2 commits a merge is easy.

**Comment by dhimmel** [comment]
> in this case if you'd leave the original commit, and squash all of our subsequent ones to a single (or possibly) multiple, logically oriented commits would be great. 

I'm going to squash all of my commits. Grouping them will be pretty burdensome and the intermediate commits were often not functional. I [created a branch](https://github.com/dhimmel/pandas/commits/compression-unsquashed) on `dhimmel/pandas` to preserve the individual commits for reference. 

Just to list things here, `git log master.. --oneline` gives the following:

640dc1b Simplify TextIOWrapper usage
4774b75 Use list to store new handles
94aa0cb Track new handles in _get_handle https://github.com/pandas-dev/pandas/pull/14576#issuecomment-265283737 https://github.com/pandas-dev/pandas/pull/14576#discussion_r91988117
376dce3 Close bz2 file object in Python 2
b88d1b8 Remove compressed URL c-engine tests
b34c2f6 Move _infer_compression to io/common.py
d27e57d Fix "ImportError: No module named backports"
97a1c25 Test compressed URLs using c engine
ecee74f Revert to old error msg. Remove unused line
7c9fe8b Skip xz test if no lzma
492e615 Remove unused imports
bf9fe97 Add invalid compression test
7940ab7 Fix PEP8 errors
05332fd Update S3 get_filepath_or_buffer compression
24ea605 Fix get_filepath_or_buffer and _get_handle
2dc1f2f Clean up docstring
10652a0 Move compression inference to io/parsers
8d24bcf Delete unused function _wrap_compressed
83b2bc5 Infer compression from URL extension
c2e6e5b Improve error message, fix partial scope
9977da2 Fix nose test generator bug. Better error messages
7ff1247 Inheret from object for TestCompressedUrl
30e1caf Fix local salaries.csv path
6230dfb Update salaries.csv URL post #14587
6530f46 Enable nose execution of test generators
a441448 Broken: use method generators to DRY tests
25114c3 Tests for bz2, xz & zip URLs
89e45f4 Revert changes to zip error messages
29b7f74 Change error message in test_zip
3aba6cb Replicate original handles appending
101a344 Impove _get_handle docstring
58ffbce Rename source to path_or_buf and flake8
f0901fb Remove ZipFile support for Python 2.6
2c78178 Improve readability
592cf92 Fix ZipFile bug in Python 3.6+
74caf72 combine compression code into one

36 commits total in this PR. Wow!

**Comment by dhimmel** [comment]
> if we only have 2 commits a merge is easy.

Great we're now at 2 commits. I think merge makes sense, because it is nice to see the diff of both commits combined. Alone the changesets will be incomplete.

Let me know if there's anything more for me to do.

**Comment by jorisvandenbossche** [comment]
@jreback You can use the github interface as well to "rebase and merge" instead of squashing, which will just apply the commits in this PR to master. And so the two commits now in the PR are perfect!

**Comment by jreback** [comment]
@jorisvandenbossche right, but this is the exception to the rule.  I will have a look in a bit. 

@dhimmel going to merge, and can followup on comments / issues in another PR.

File: pandas/io/common.py (hunk 20 above)
Comment (jorisvandenbossche) [inline pandas/io/common.py]: @dhimmel This looks like the appropriate approach!

**Comment by jreback** [comment]
thanks @dhimmel very nice!

in your xref PR, pls add whatsnew that cover all of the issues that were closed (for 0.20.0)

**Comment by jorisvandenbossche** [comment]
@dhimmel thanks a lot!
